dbms/src/AggregateFunctions/AggregateFunctionCategoricalInformationValue.h

```diff
@@ -0,0 +1,156 @@
+#pragma once
+
+#include <AggregateFunctions/IAggregateFunction.h>
+#include <Columns/ColumnArray.h>
+#include <Columns/ColumnsNumber.h>
+#include <DataTypes/DataTypeArray.h>
+#include <DataTypes/DataTypesNumber.h>
+#include <IO/ReadHelpers.h>
+#include <IO/WriteHelpers.h>
+
+#include <ext/range.h>
+
+
+namespace DB
+{
+
+template <typename T = UInt64>
+class AggregateFunctionCategoricalIV final : public IAggregateFunctionHelper<AggregateFunctionCategoricalIV<T>>
+{
+private:
+    size_t category_count;
+
+public:
+    AggregateFunctionCategoricalIV(const DataTypes & arguments_, const Array & params_) :
+        IAggregateFunctionHelper<AggregateFunctionCategoricalIV<T>> {arguments_, params_},
+        category_count {arguments_.size() - 1}
+    {
+        // notice: argument types has been checked before
+    }
+
+    String getName() const override
+    {
+        return "categoricalInformationValue";
+    }
+
+    const char * getHeaderFilePath() const override
+    {
+        return __FILE__;
+    }
+
+    void create(AggregateDataPtr place) const override
+    {
+        memset(place, 0, sizeOfData());
+    }
+
+    void destroy(AggregateDataPtr) const noexcept override
+    {
+        // nothing
+    }
+
+    bool hasTrivialDestructor() const override
+    {
+        return true;
+    }
+
+    size_t sizeOfData() const override
+    {
+        return sizeof(T) * (category_count + 1) * 2;
+    }
+
+    size_t alignOfData() const override
+    {
+        return alignof(T);
+    }
+
+    void add(
+        AggregateDataPtr place,
+        const IColumn ** columns,
+        size_t row_num,
+        Arena *
+    ) const override
+    {
+        auto y_col = static_cast<const ColumnUInt8 *>(columns[category_count]);
+        bool y = y_col->getData()[row_num];
+
+        for (size_t i : ext::range(0, category_count))
+        {
+            auto x_col = static_cast<const ColumnUInt8 *>(columns[i]);
+            bool x = x_col->getData()[row_num];
+
+            if (x)
+                reinterpret_cast<T *>(place)[i * 2 + size_t(y)] += 1;
```

Comment:
But we have `IAggregateFunctionDataHelper` that allows to avoid reinterpret_cast. Why don't use it?